### ISSUE TITLE
eval ppl problem

### ISSUE THREAD

**Issue by SUSTechBruce** [facebookresearch/llama-recipes#82]
### System Info

Pytorch version > 2.0
GPUs: A100 80G

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### 🐛 Describe the bug

Does anyone know why the eval ppl is much higher than train ppl when I want to run:
```
torchrun --nnodes 1 --nproc_per_node 4  llama_finetuning.py \
 --enable_fsdp \
 --use_peft \
 --peft_method lora \
 --model_name meta-llama/Llama-2-7b-hf  --pure_bf16 \
 --output_dir path_to_sft_checkpoint \
 --dataset samsum_dataset \
 --batch_size_training 8 \
 --lr 2e-4
```
And the results are:
```
eval_ppl=tensor(975.5518, device='cuda:2') eval_epoch_loss=tensor(6.8830, device='cuda:2') eval_ppl=tensor(975.5518, device='cuda:3') eval_epoch_loss=tensor(6.8830, device='cuda:3')

 eval_ppl=tensor(975.5518, device='cuda:0') eval_epoch_loss=tensor(6.8830, device='cuda:0')
we are in the saving the PEFT modules
we are in the saving the PEFT modules
we are in the saving the PEFT modules
 eval_ppl=tensor(975.5518, device='cuda:1') eval_epoch_loss=tensor(6.8830, device='cuda:1')
we are in the saving the PEFT modules
PEFT modules are saved in path_to_sft_checkpoint directory
PEFT modules are saved in path_to_sft_checkpoint directory
Epoch 1: train_perplexity=1.5734, train_epoch_loss=0.4533
Epoch 1: train_perplexity=1.5734, train_epoch_loss=0.4533
PEFT modules are saved in path_to_sft_checkpoint directory
Epoch 1: train_perplexity=1.5734, train_epoch_loss=0.4533
PEFT modules are saved in path_to_sft_checkpoint directory
best eval loss on epoch 0 is 6.883003234863281
Epoch 1: train_perplexity=1.5734, train_epoch_loss=0.4533
```
The eval_ppl is 975.5518 and the train_ppl is 1.5734, the gap is large.

### Error logs

```
eval_ppl=tensor(975.5518, device='cuda:2') eval_epoch_loss=tensor(6.8830, device='cuda:2') eval_ppl=tensor(975.5518, device='cuda:3') eval_epoch_loss=tensor(6.8830, device='cuda:3')

 eval_ppl=tensor(975.5518, device='cuda:0') eval_epoch_loss=tensor(6.8830, device='cuda:0')
```

### Expected behavior

Why

**Comment by HamidShojanazeri** [comment]
@SUSTechBruce wondering if you are on the latest code. haven't got the chance to run this setting yet but with the default settings, bs=4, and default lr in my logs I see the following. Will try to repro it, BTW pls make sure on PyTorch nightlies as using FSDP + PEFT.

```
Key: avg_train_prep, Value: 4.719568252563477
Key: avg_train_loss, Value: 1.504786491394043
Key: avg_eval_prep, Value: 5.776612758636475
Key: avg_eval_loss, Value: 1.6824092864990234
```

**Comment by HamidShojanazeri** [comment]
I used tried to repro with 8xA100 40 GB.

```
 torchrun --nnodes 1 --nproc_per_node 8  llama_finetuning.py --enable_fsdp --use_peft --peft_method lora --model_name meta-llama/Llama-2-7b-chat-hf --output_dir ./fsdp-models-rabk0 --num_epochs 2 --pure_bf16 --lr 2e-4 --batch_size_training 8
 ```
 
 still seeing the results
 
 ```
  eval_ppl=tensor(6.5052, device='cuda:0') eval_epoch_loss=tensor(1.8726, device='cuda:0')
we are about to save the PEFT modules
PEFT modules are saved in ./fsdp-models-rabk0 directory
best eval loss on epoch 1 is 1.8725948333740234
Epoch 2: train_perplexity=2.5945, train_epoch_loss=0.9534
Key: avg_train_prep, Value: 2.7766218185424805
Key: avg_train_loss, Value: 1.0190796852111816
Key: avg_eval_prep, Value: 6.802661895751953
Key: avg_eval_loss, Value: 1.9163568019866943
```

**Comment by SUSTechBruce** [comment]
thanks! I will try it again.



---Original---
From: "Hamid ***@***.***&gt;
Date: Thu, Aug 3, 2023 01:00 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [facebookresearch/llama-recipes] eval ppl problem (Issue #82)




 
I used tried to repro with 8xA100 40 GB.
  torchrun --nnodes 1 --nproc_per_node 8  llama_finetuning.py --enable_fsdp --use_peft --peft_method lora --model_name meta-llama/Llama-2-7b-chat-hf --output_dir ./fsdp-models-rabk0 --num_epochs 2 --pure_bf16 --lr 2e-4 --batch_size_training 8  
still seeing the results
  eval_ppl=tensor(6.5052, device='cuda:0') eval_epoch_loss=tensor(1.8726, device='cuda:0') we are about to save the PEFT modules PEFT modules are saved in ./fsdp-models-rabk0 directory best eval loss on epoch 1 is 1.8725948333740234 Epoch 2: train_perplexity=2.5945, train_epoch_loss=0.9534 Key: avg_train_prep, Value: 2.7766218185424805 Key: avg_train_loss, Value: 1.0190796852111816 Key: avg_eval_prep, Value: 6.802661895751953 Key: avg_eval_loss, Value: 1.9163568019866943

**Comment by HamidShojanazeri** [comment]
@SUSTechBruce , sure, I tried to repro with 4xA100 40 GB.

```
 torchrun --nnodes 1 --nproc_per_node 4  llama_finetuning.py --enable_fsdp --use_peft --peft_method lora --model_name meta-llama/Llama-2-7b-chat-hf --output_dir ./fsdp-models-rabk0 --num_epochs 2 --pure_bf16 --lr 2e-4 --batch_size_training 8
 ```
 
 still seeing the results
 
 ```
we are about to save the PEFT modules
PEFT modules are saved in ./fsdp-models-rabk0 directory
best eval loss on epoch 1 is 1.7735178470611572
Epoch 2: train_perplexity=2.4605, train_epoch_loss=0.9004
Key: avg_train_prep, Value: 2.614932060241699
Key: avg_train_loss, Value: 0.9594911932945251
Key: avg_eval_prep, Value: 6.174006938934326
Key: avg_eval_loss, Value: 1.8193004131317139
```

**Comment by SUSTechBruce** [comment]
Hi bro, can you try alpaca dataset too? I use default setting and get 45.7 eval ppl.&nbsp;



---Original---
From: "Hamid ***@***.***&gt;
Date: Thu, Aug 3, 2023 01:26 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [facebookresearch/llama-recipes] eval ppl problem (Issue #82)




 
@SUSTechBruce , sure, I tried to repro with 4xA100 40 GB.
  torchrun --nnodes 1 --nproc_per_node 4  llama_finetuning.py --enable_fsdp --use_peft --peft_method lora --model_name meta-llama/Llama-2-7b-chat-hf --output_dir ./fsdp-models-rabk0 --num_epochs 2 --pure_bf16 --lr 2e-4 --batch_size_training 8  
still seeing the results
 we are about to save the PEFT modules PEFT modules are saved in ./fsdp-models-rabk0 directory best eval loss on epoch 1 is 1.7735178470611572 Epoch 2: train_perplexity=2.4605, train_epoch_loss=0.9004 Key: avg_train_prep, Value: 2.614932060241699 Key: avg_train_loss, Value: 0.9594911932945251 Key: avg_eval_prep, Value: 6.174006938934326 Key: avg_eval_loss, Value: 1.8193004131317139

**Comment by HamidShojanazeri** [comment]
addressed in this[ PR](https://github.com/facebookresearch/llama-recipes/pull/107).
